### ISSUE TITLE
gtest-port.h fileno not declared in scope

### ISSUE THREAD

**Issue by ssi112** [google/googletest#813]
I'm getting errors in gtest-port.h. Have tried various flags none are working. 

Have searched and found similar errors, but no definitive solutions.

Any help is appreciated.

Thanks!

CYGWIN_NT-6.1 BaguaDragon 2.5.2(0.297/5/3) 2016-06-23 14:29 x86_64 Cygwin
g++ (GCC) 5.4.0

g++ **-static -std=c++0x -DGTEST_OS_CYGWIN=1** -pthread -MMD -c main.cpp -o build/main.o -I /home/ssi112/cpp/googletest-master/googletest -I /home/ssi112/cpp/googletest-master/googletest/include -I . -I ../src -I ../../class-library/include
In file included from /home/ssi112/cpp/googletest-master/googletest/include/gtest/internal/gtest-internal.h:40:0,
                 from /home/ssi112/cpp/googletest-master/googletest/include/gtest/gtest.h:58,
                 from main.cpp:9:
/home/ssi112/cpp/googletest-master/googletest/include/gtest/internal/gtest-port.h: In function ‘int testing::internal::posix::FileNo(FILE_)’:
/home/ssi112/cpp/googletest-master/googletest/include/gtest/internal/gtest-port.h:2340:51: error: ‘fileno’ was not declared in this scope
 inline int FileNo(FILE_ file) { return fileno(file); }
                                                   ^
/home/ssi112/cpp/googletest-master/googletest/include/gtest/internal/gtest-port.h: In function ‘char\* testing::internal::posix::StrDup(const char_)’:
/home/ssi112/cpp/googletest-master/googletest/include/gtest/internal/gtest-port.h:2346:57: error: ‘strdup’ was not declared in this scope
 inline char_ StrDup(const char\* src) { return strdup(src); }
                                                         ^
/home/ssi112/cpp/googletest-master/googletest/include/gtest/internal/gtest-port.h: In function ‘FILE\* testing::internal::posix::FDOpen(int, const char_)’:
/home/ssi112/cpp/googletest-master/googletest/include/gtest/internal/gtest-port.h:2374:71: error: ‘fdopen’ was not declared in this scope
 inline FILE_ FDOpen(int fd, const char\* mode) { return fdopen(fd, mode); }
                                                                       ^
In file included from main.cpp:12:0:
aggregate_test.hpp: In member function ‘virtual void AggregateTest_verifySplitAndFindAggregateFunctionality_Test::TestBody()’:
aggregate_test.hpp:11:32: error: conversion from ‘void’ to non-scalar type ‘std::string {aka std::basic_string<char>}’ requested
  std::string actual = aggregate(1, line);
                                ^
make: **\* [Makefile:28: build/main.o] Error 1

**Comment by jiangzzz** [comment]
Same error for me using Cygwin on Windows 10.
Everything works fine when working on Ubuntu with the same source files...
Everything is updated to the current last version: Cygwin, Windows 10 and Ubuntu.

**Comment by thomasbenard** [comment]
I can confirm. I have the exact same issue with Cygwin on Windows 10

**Comment by dneto0** [comment]
I have a similar situation.  Compilation fails like this with -std=c++11 but then passes with -std=gnu++11

**Comment by jfinkhaeuser** [comment]
This issue isn't exactly solved. Any project that has stricter requirements, i.e. for `c++xx` instead of `gnu++xx`, gtest still doesn't work.

**Comment by jfinkhaeuser** [comment]
I should add, it also blatantly goes against the README requirement of a `c++11` compliant compiler. Clearly that's not enough :)

**Comment by gennadiycivil** [comment]
Anything related to CMake is community supported. We do not use it internally and we are not in position to check. In general the maintainers would look for community consensus. Someone needs to review such PRs and comment on them.
With that, should I re-open this issue? Please let me know
Thanks

**Comment by jfinkhaeuser** [comment]
Yes please! It's got nothing to do with CMake from where I'm standing.

**Comment by gennadiycivil** [comment]
@jfinkhaeuser I am re-opening this issue. It is not clear from the description and the conversation exactly what the problem is and on what platform. If this is CygWin specific - the maintainers are not in position to help. We do not use CygWin and we do not have any tests running against it. it would have to be a community effort to troubleshoot and support CygWin.

**Comment by gennadiycivil** [comment]
The best possible outcome would be 
1) Find the solution
2) Adding CI configuration ( I guess on appveyor.xml since cygwin runs under windows ) to make sure that the problem stays fixed going forward

**Comment by jfinkhaeuser** [comment]
The problem is that on Cygwin at least, the `gtest-ports.h` header is looking for e.g. the `fileno` function, which under the C++11 standard does not exist. The extended GNU++11 standard does support it.

So the README's claim that a C++11 compliant compiler is required isn't true on Cygwin; Cygwin requires more from the compiler.

I ran into the issue on Cygwin, but really that's a more general problem of what C++ standard you actually support - it just happens to break on Cygwin.

**Comment by vslashg** [comment]
Revisiting this bug: we don't support Cygwin, and aren't in a position to test and fix this.

If the community can provide a patch which fixes this in Cygwin and does not impact supported platforms, we would be happy to accept it.

**Comment by jfinkhaeuser** [comment]
Let me see if I can find time for this today.

**Comment by jfinkhaeuser** [comment]
Reproduce:

```bash
$ cd /path/to/build
$ CXXFLAGS=-std=c++11 cmake /path/to/source
$ make
```

Will open a PR later.

**Comment by jfinkhaeuser** [comment]
Workaround:

```bash
$ cd /path/to/build
$ CXXFLAGS=-std=gnu++11 cmake /path/to/source
$ make
```

TL;DR is, some functions aren't standard, but part of GNU.

**Comment by rohitmanokaran** [comment]
> Reproduce:
> 
> ```shell
> $ cd /path/to/build
> $ CXXFLAGS=-std=c++11 cmake /path/to/source
> $ make
> ```
> 
> Will open a PR later.

Running CXXFLAGS=-std=c++11 cmake . && make seems to be working for me. Has this been solved?

**Comment by automorphis** [comment]
> I have no idea if something was changed, to be honest. It's been a while since anyone commented here. I've also not got a cygwin installation any longer at the moment.

I have this issue as well on Cygwin.

**Comment by automorphis** [comment]
I was able to successfully build and run on Cygwin by appending the line

`set(CMAKE_CXX_EXTENSIONS ON)`

to the CMake file. Incidentally, another Google open-source project (LevelDB) had the same problem, with the same solution.

**Comment by jfinkhaeuser** [comment]
Look, the problem isn't building and running googletest by itself. It's building and running downstream projects that want to include googletest headers.

TL;DR is, the README claims C++11 is all that's required, but the fact is, that a POSIX API is also required. This is the case on e.g. Linux, but not on Cygwin - not by default.

The linked PR adjusts the README accordingly.

**Comment by automorphis** [comment]
Yes, I see your point. I just wanted to suggest a solution for other people in the same situation as me, who just need the code to build and run. I am not very fluent in C++ and it took me nearly an entire workday to figure out the solution.

> Look, the problem isn't building and running googletest by itself. It's building and running downstream projects that want to include googletest headers.
> 
> TL;DR is, the README claims C++11 is all that's required, but the fact is, that a POSIX API is also required. This is the case on e.g. Linux, but not on Cygwin - not by default.
> 
> The linked PR adjusts the README accordingly.
